frontend/packages/console-shared/src/components/dynamic-form/utils.ts

```diff
@@ -54,30 +54,100 @@ export const hasNoFields = (jsonSchema: JSONSchema6 = {}, uiSchema: UiSchema = {
+ * Fields are weighted according to the following tiers:
+ *  Tier 1 (negative 10^9 - 10^6 magnitude):  Required fields with ui schema
+ *  Tier 2 (negative 10^9 magnitude):         Required fields without ui schema
+ *  Tier 3 (negative 10^6 magnitude):         Optional fields with ui schema
+ *  Tier 4 (positive 10^3 maginitude):        Control fields that don't fit any above
+ *  Tier 5 (Infinity):                        All other fields
+ *
+ * Within each of the above tiers, fields are further weighted based on field dependency and ui
+ * schema defined sort order:
+ *   - Fields without dependency - base weight + ui schema sort order
+ *   - Control field - base weight + ui schema sort order + (nth control field) * 10000
+ *   - Dependent field - control field weight + ui schema sort order + 1
+ *
+ * These weight numbers are arbitrary, but spaced far enough apart to prevent collisions.
+ */
+const getJSONSchemaPropertySortWeight = (
+  property: string,
+  jsonSchema: JSONSchema6,
+  uiSchema: UiSchema,
+): number => {
+  const isRequired = (jsonSchema?.required ?? []).includes(property);
+  const propertyUISchema = uiSchema?.[property];
+
+  // Any sibling has a dependency with this as the control field.
+  const isControlField = _.some(
+    uiSchema,
+    ({ 'ui:dependency': dependency }) => dependency?.controlFieldName === property,
+  );
+
+  // Property sort order from ui schema (adjusted to 1-based origin). Use propertyNames.length
+  // when no uiSchema sortOrder exists. Ensures properties without uiSchema sort order  have a
+  // higher weight than those with.
+  const propertySortOrder = Number(
+    propertyUISchema?.['ui:sortOrder'] ?? _.keys(jsonSchema?.properties).length,
+  );
+
+  // This property's control field name, if it exists
+  const controlFieldName = propertyUISchema?.['ui:dependency']?.controlFieldName;
+
+  // A small offset that is added to the base weight so that control fields get sorted
+  // below other fields in the same 'tier', and allows for depenendt fields to be sorted
+  // directly after their control field.
+  const controlFieldOffset = isControlField ? propertySortOrder * THOUSAND : 0;
+
+  // Total offset to be added to base tier
+  const offset = controlFieldOffset + propertySortOrder;
+
+  // If this property is a dependent, it's weight is based on it's control field
+  if (controlFieldName) {
+    return getJSONSchemaPropertySortWeight(controlFieldName, jsonSchema, uiSchema) + offset + 1;
+  }
+
+  // Tier 1 = -1001000000 (negative one billion one million) + offset
+  // Tier 2 = -1000000000 (negagive one billion) + offset
+  // Tier 3 = -1000000 (negative one million) + offset
+  // Tier 4 = 0 + offset
+  // Tier 5 = Infinity
+  return (
+    // Doesn't meet any sorting criteria, set to infinity
+    (!isRequired && !propertyUISchema && !controlFieldOffset ? Infinity : 0) -
+    (isRequired ? BILLION : 0) -
+    (propertyUISchema ? MILLION : 0) +
+    offset
+  );
+};
```

Comment:
Pulled this function out of `getJSONSchemaOrder` scope.

Reply:
I tried to improve the comments and logic to make this a little bit easier to understand.